File: interpreter.py
import random
from dataclasses import dataclass, field
from enum import Enum
from typing import Union, get_args

class Immediate(int):
    """An immediate/literal value."""
    pass

class MemoryAddress(int):
    """A memory address."""
    pass

class LineNumber(int):
    """A line number."""
    pass


@dataclass
class Instruction:
    pass

class Register(Enum):
    R0 = 0
    R1 = 1
    R2 = 2
    R3 = 3

@dataclass
class Machine:
    instructions: list[str]
    code: list[Instruction]
    registers: dict[Register, int] = field(default_factory=lambda: {r: random.randint(0, 255) for r in Register})
    memory: list[int] = field(default_factory=lambda: [random.randint(0, 255) for _ in range(64)])
    pc: int = 0
    halted: bool = False

    def load_instructions_from_file(self, filename: str):
        with open(filename, "r") as f:
            self.instructions = f.readlines()
        self.instructions = [*filter(lambda l: l.strip() and not l.strip().startswith(";"), self.instructions)]
        self.code = [*enumerate(map(self.parse_instruction, self.instructions))]


    def parse_instruction(self, line) -> Instruction:
        Token_types = {
            "r": Register,
            "#": Immediate,
            "@": MemoryAddress,
            "l": LineNumber
        }
        tokens = [*filter(None, map(lambda x: x.strip(), line.split(" ")))]
        try:
            opcode_info = Inst[tokens[0]]
        except ValueError:
            raise ValueError(f"Unknown opcode: {tokens[0]}")
        opcode_args = Inst[tokens[0]]["args"]
        given_args = tokens[1:]
        try:
            assert(len(opcode_args) == len(given_args))
        except AssertionError:
            raise ValueError(
                f"Expected {len(opcode_args)} arguments for {tokens[0]}, got {len(given_args)}"
            )
        try:
            given_args_types = [Token_types[operand[0]] for operand in given_args]
            for i, (expected, actual) in enumerate(zip(opcode_args, given_args_types)):
                if hasattr(expected, '__origin__') and expected.__origin__ is Union:
                    if actual not in get_args(expected):
                        raise ValueError(
                            f"Argument {i+1} for {tokens[0]}: expected one of {get_args(expected)}, got {actual}"
                        )
                else:
                    if actual is not expected:
                        raise ValueError(
                            f"Argument {i+1} for {tokens[0]}: expected {expected}, got {actual}"
                        )
        except KeyError as e:
            raise ValueError(f"Unknown token prefix: {e}")
        given_args_values = [t(int(operand[1:])) for t, operand in zip(given_args_types, given_args)]
        return  opcode_info["inst"](*given_args_values)
        
    def print_code(self):
        for idx, instr in self.code:
            marker = " <== PC" if idx == self.pc else ""
            print(f"{idx}: {instr}{marker}")

    def print_memory(self):
        for i in range(0, 64, 8):
            print([f"{x:02x}" for x in self.memory[i:i+8]], sep=" ")

    def print_registers(self):
        print("Registers:")
        for reg, val in self.registers.items():
            print(f"r{reg.value}: {val}")

    def execute(self):
        index = self.pc
        self.pc += 1
        self.code[index][1].execute(self)

    def run(self, max_steps=100000):
        steps = 0
        while self.pc < len(self.code) and not self.halted and steps < max_steps:
            self.execute()
            steps += 1
        return self.halted

    def run_step_by_step(self):
        while self.pc < len(self.code) and not self.halted:
            self.print_code()
            self.print_registers()
            self.print_memory()
            input("Press Enter to execute next instruction...")
            idx, instr = self.code[self.pc]
            self.execute()
            print("-" * 40)
        self.print_code()
        self.print_registers()
        print(f"\nProgram halted. r0 = {self.registers[Register.R0]}")


@dataclass
class SET(Instruction):
    """SET dest, src — dest and src can be Reg, Mem, or (src only) Immediate."""
    dest: Union[Register, MemoryAddress]
    src: Union[Register, MemoryAddress, Immediate]

    def _get_value(self, machine: Machine, operand: Union[Register, MemoryAddress, Immediate]) -> int:
        if isinstance(operand, Register):
            return machine.registers[operand]
        if isinstance(operand, MemoryAddress):
            return machine.memory[operand]
        return int(operand)  # Immediate

    def execute(self, machine: Machine) -> None:
        value = self._get_value(machine, self.src)
        if isinstance(self.dest, Register):
            machine.registers[self.dest] = value
        else:
            machine.memory[self.dest] = value

@dataclass
class INC(Instruction):
    x: Register

    def execute(self, machine: Machine) -> None:
        machine.registers[self.x] += 1

@dataclass
class DEC(Instruction):
    x: Register

    def execute(self, machine: Machine) -> None:
        machine.registers[self.x] -= 1

@dataclass
class ISZ(Instruction):
    x: Register

    def execute(self, machine: Machine) -> None:
        if machine.registers[self.x] == 0:
            machine.pc += 1

@dataclass
class ISN(Instruction):
    """Skip next instruction if register is negative."""
    x: Register

    def execute(self, machine: Machine) -> None:
        if machine.registers[self.x] < 0:
            machine.pc += 1

class STP(Instruction):
    def execute(self, machine: Machine) -> None:
        machine.halted = True

@dataclass
class JMP(Instruction):
    x: LineNumber

    def execute(self, machine: Machine) -> None:
        machine.pc = self.x - 1  # source line numbers are 1-based (l1=first, l2=second, …)


Inst = {
    "SET": {
        "args": [Union[Register, MemoryAddress], Union[Register, MemoryAddress, Immediate]],
        "inst": SET
    },
    "INC": {
        "args": [Register],
        "inst": INC
    },
    "DEC": {
        "args": [Register],
        "inst": DEC
    },
    "ISZ": {
        "args": [Register],
        "inst": ISZ
    },
    "ISN": {
        "args": [Register],
        "inst": ISN
    },
    "STP": {
        "args": [],
        "inst": STP
    },
    "JMP": {
        "args": [LineNumber],
        "inst": JMP
    },
}

if __name__ == "__main__":
    m = Machine(instructions=[], code=[])
    m.load_instructions_from_file("code.x")
    m.run_step_by_step()

File: web/interpreter.js
/**
 * MiniASM VM — vanilla JavaScript implementation.
 * Instruction set: SET (Reg/Mem, Reg/Mem/Immediate), INC, DEC, ISZ, ISN, STP, JMP.
 */

const Register = { R0: 0, R1: 1, R2: 2, R3: 3 };
const REG_NAMES = ['r0', 'r1', 'r2', 'r3'];

function parseOperand(token) {
  if (!token || token.length < 2) throw new Error(`Invalid token: ${token}`);
  const prefix = token[0];
  const value = parseInt(token.slice(1), 10);
  if (Number.isNaN(value)) throw new Error(`Invalid number in token: ${token}`);
  switch (prefix) {
    case 'r': return { type: 'Register', value };
    case '#': return { type: 'Immediate', value };
    case '@': return { type: 'MemoryAddress', value };
    case 'l': return { type: 'InstructionNumber', value };
    default: throw new Error(`Unknown token prefix: ${prefix}`);
  }
}

const INST = {
  SET: {
    args: [
      ['Register', 'MemoryAddress'],
      ['Register', 'MemoryAddress', 'Immediate']
    ],
    run(m, dest, src) {
      const v = getValue(m, src);
      setValue(m, dest, v);
    }
  },
  INC: {
    args: [['Register']],
    run(m, x) { m.registers[x.value] += 1; }
  },
  DEC: {
    args: [['Register']],
    run(m, x) { m.registers[x.value] -= 1; }
  },
  ISZ: {
    args: [['Register']],
    run(m, x) { if (m.registers[x.value] === 0) m.pc += 1; }
  },
  ISN: {
    args: [['Register']],
    run(m, x) { if (m.registers[x.value] < 0) m.pc += 1; }
  },
  ADD: {
    args: [['Register'], ['Register']],
    run(m, x, y) { m.registers[x.value] += m.registers[y.value]; }
  },
  MUL: {
    args: [['Register'], ['Register']],
    run(m, x, y) { m.registers[x.value] *= m.registers[y.value]; }
  },
  POW: {
    args: [['Register'], ['Register']],
    run(m, x, y) { m.registers[x.value] = Math.pow(m.registers[x.value], m.registers[y.value]); }
  },
  STP: {
    args: [],
    run(m) { m.halted = true; }
  },
  JMP: {
    args: [['InstructionNumber']],
    run(m, line) { m.pc = line.value - 1; }  // source line numbers are 1-based (l1=first, l2=second, …)
  }
};

function checkArg(opcode, argIndex, operand, allowedTypes) {
  if (!allowedTypes.includes(operand.type)) {
    throw new Error(
      `Argument ${argIndex + 1} for ${opcode}: expected one of [${allowedTypes.join(', ')}], got ${operand.type}`
    );
  }
}

function getValue(m, op) {
  switch (op.type) {
    case 'Register': return m.registers[op.value];
    case 'MemoryAddress': return m.memory[op.value];
    case 'Immediate': return op.value;
    default: throw new Error(`Cannot get value for type: ${op.type}`);
  }
}

function setValue(m, dest, value) {
  switch (dest.type) {
    case 'Register': m.registers[dest.value] = value; break;
    case 'MemoryAddress': m.memory[dest.value] = value; break;
    default: throw new Error(`Cannot set value for type: ${dest.type}`);
  }
}

function parseInstruction(line) {
  const tokens = line.split(/\s+/).map(s => s.trim()).filter(Boolean);
  if (tokens.length === 0) return null;
  const opcode = tokens[0];
  const info = INST[opcode];
  if (!info) throw new Error(`Unknown opcode: ${opcode}`);
  const givenArgs = tokens.slice(1);
  if (info.args.length !== givenArgs.length) {
    throw new Error(`Expected ${info.args.length} arguments for ${opcode}, got ${givenArgs.length}`);
  }
  const operands = givenArgs.map(t => parseOperand(t));
  for (let i = 0; i < info.args.length; i++) {
    checkArg(opcode, i, operands[i], info.args[i]);
  }
  return { opcode, operands };
}

function createMachine() {
  const rand = () => Math.floor(Math.random() * 256);
  return {
    instructions: [],
    code: [],
    registers: { [Register.R0]: rand(), [Register.R1]: rand(), [Register.R2]: rand(), [Register.R3]: rand() },
    memory: Array.from({ length: 64 }, rand),
    pc: 0,
    halted: false
  };
}

/**
 * Load program into machine. Accepts:
 * - string: raw source (e.g. from textarea)
 * - object: { source: string } or { lines: string[] }
 */
function loadProgram(machine, program) {
  let text;
  if (typeof program === 'string') {
    text = program;
  } else if (program && typeof program === 'object') {
    if (typeof program.source === 'string') text = program.source;
    else if (Array.isArray(program.lines)) text = program.lines.join('\n');
    else throw new Error('Program object must have "source" or "lines"');
  } else {
    throw new Error('Program must be a string or { source | lines } object');
  }
  const allLines = text.split(/\n/).map(l => l.trim());
  const lines = [];
  const lineNumbers = [];
  for (let i = 0; i < allLines.length; i++) {
    const line = allLines[i];
    if (line && !line.startsWith(';')) {
      lines.push(line);
      lineNumbers.push(i + 1);
    }
  }
  machine.instructions = lines;
  machine.code = lines.map((line, idx) => [idx, parseInstruction(line)]);
  machine.lineNumbers = lineNumbers;
  return machine;
}

function execute(machine) {
  const [index, instr] = machine.code[machine.pc];
  INST[instr.opcode].run(machine, ...instr.operands);
  if (machine.halted) return;
  if (instr.opcode !== 'JMP') {
    machine.pc += 1;
  }
}

function run(machine, maxSteps = 100000) {
  let steps = 0;
  while (machine.pc < machine.code.length && !machine.halted && steps < maxSteps) {
    execute(machine);
    steps += 1;
  }
  return machine.halted;
}

function printCode(machine) {
  for (const [idx, instr] of machine.code) {
    const marker = idx === machine.pc ? ' <== PC' : '';
    const argsStr = instr.operands.map(op => {
      if (op.type === 'Register') return REG_NAMES[op.value];
      if (op.type === 'Immediate') return '#' + op.value;
      if (op.type === 'MemoryAddress') return '@' + op.value;
      return 'l' + op.value;
    }).join(' ');
    console.log(`${idx}: ${instr.opcode} ${argsStr}${marker}`);
  }
}

function printRegisters(machine) {
  console.log('Registers:');
  for (let i = 0; i <= 3; i++) {
    console.log(`r${i}: ${machine.registers[i]}`);
  }
}

function printMemory(machine) {
  for (let i = 0; i < 64; i += 8) {
    const row = machine.memory.slice(i, i + 8).map(x => x.toString(16).padStart(2, '0'));
    console.log(row.join(' '));
  }
}

// Export for use as module or in browser (code from DOM via loadProgram(m, { source: el.value }))
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    Register,
    createMachine,
    loadProgram,
    execute,
    run,
    parseInstruction,
    printCode,
    printRegisters,
    printMemory,
    INST
  };
}
if (typeof window !== 'undefined') {
  window.MiniASM = {
    Register,
    createMachine,
    loadProgram,
    execute,
    run,
    parseInstruction,
    printCode,
    printRegisters,
    printMemory,
    INST
  };
}
